File: Shapes/Wave/WebGL/Scene.js
﻿// Définition de la classe Scene

// superclasses et classes nécessaires
Requires("libs/Misc/TrackBallScene.js");
Requires("libs/Material/DeferredShadingMaterial.js");
Requires("libs/Material/OmniLight.js");
Requires("libs/Misc/MeshStructureDraw.js");
Requires("Wave");


class Scene extends TrackBallScene
{
    /**
     * Constructeur
     */
    constructor()
    {
        super(true);

        // créer le matériau de l'objet
        let Kd = vec4.fromValues(0.2, 0.4, 1.0, 1.0);
        let Ks = vec3.fromValues(0.9, 0.9, 0.9);
        let Ns = 128;
        this.m_Material = new DeferredShadingMaterial(Kd, Ks, Ns);

        // définir une lampe directionnelle
        this.m_Light0 = new OmniLight();
        this.m_Light0.setPosition(vec4.fromValues(2, 3, 2, 0));
        this.m_Light0.setColor(vec3.fromValues(0.8,0.8,0.8));
        this.addLight(this.m_Light0);

        // définir une lampe positionnelle
        this.m_Light1 = new OmniLight();
        this.m_Light1.setPosition(vec4.fromValues(-10, 5, 5, 1));
        this.m_Light1.setColor(vec3.fromValues(150,150,150));
        this.addLight(this.m_Light1);

        // définir une lampe directionnelle
        this.m_Light2 = new OmniLight();
        this.m_Light2.setPosition(vec4.fromValues(-2, 1, -2, 0));
        this.m_Light2.setColor(vec3.fromValues(0.4,0.4,0.4));
        this.addLight(this.m_Light2);

        // créer le contour de la fonction
        let segments_count = 96;
        let border = [];
        for (let segment=segments_count; segment>=0; segment--) {
            // aller de l'extérieur vers l'intérieur
            let r = 40.0 * segment/segments_count;
            // calculer y = f(r)
            let y = -5.0;
            if (segment > 0) {
                y = -5.0 * Math.sin(r) / (r*r);
            }
            r = r / 10.0;
            border.push(vec2.fromValues(r,y));
        }

        // créer l'objet
        this.m_Object = new Wave(border, 48, segments_count);

        // configurer les modes de dessin
        gl.enable(gl.DEPTH_TEST);
        gl.enable(gl.CULL_FACE);

        // couleur du fond
        gl.clearColor(0.4, 0.4, 0.4, 0.0);

        // paramètres de la vue
        this.m_CameraDistance = 10;
        quat.setAxisAngle(this.m_QuatP1P2, vec3.fromValues(1,0,0), Utils.radians(30.0));
        quat.multiply(this.m_QuatOrientation, this.m_QuatP1P2, this.m_QuatOrientation);

        // modes de dessin
        this.m_MeshStructureDraw = new MeshStructureDraw(this.m_Object.getMesh(), this.m_Material, null, null, null, 0.1);
    }


    /**
     * appelée quand la taille de la vue OpenGL change
     * @param width : largeur en nombre de pixels de la fenêtre
     * @param height : hauteur en nombre de pixels de la fenêtre
     */
    onSurfaceChanged(width, height)
    {
        // appeler la méthode de la superclasse
        super.onSurfaceChanged(width, height);

        // matrice de projection (champ de vision)
        mat4.perspective(this.m_Mat4Projection, Utils.radians(15.0), width / height, 0.1, 20.0);
    }


    /**
     * appelée quand on appuie sur une touche du clavier
     * @param code : touche enfoncée
     */
    onKeyDown(code)
    {
        switch (code) {
        case 'T'.charCodeAt(0):
            this.m_MeshStructureDraw.nextTrianglesMode();
            break;
        case 'Y'.charCodeAt(0):
            this.m_MeshStructureDraw.nextEdgesMode();
            break;
        case 'U'.charCodeAt(0):
            this.m_MeshStructureDraw.nextNormalsMode();
            break;
        default:
            // appeler la méthode de la superclasse
            super.onKeyDown(code);
        }
    }


    /**
     * dessine les objets de la scène
     * @param mat4Projection : matrice de projection
     * @param mat4ModelView : matrice de vue
     */
    onDraw(mat4Projection, mat4ModelView)
    {
        // dessiner l'objet selon les modes demandés
        this.m_MeshStructureDraw.onDraw(mat4Projection, mat4ModelView);
    }


    /**
     * Cette méthode supprime les ressources allouées
     */
    destroy()
    {
        this.m_MeshStructureDraw.destroy();
        this.m_Object.destroy();
        this.m_Light2.destroy();
        this.m_Light1.destroy();
        this.m_Light0.destroy();
        this.m_Material.destroy();
        super.destroy();
    }
}

File: Transforms/Skeleton/Android/app/src/main/java/com/dbooker/livreopengl/Scene.java
package com.dbooker.livreopengl;

import static android.opengl.GLES20.*;

import com.dbooker.livreopengl.mesh.MeshObject;
import com.dbooker.livreopengl.mesh.MeshObjectFromObj;
import com.dbooker.livreopengl.misc.TurnTableScene;
import com.dbooker.livreopengl.glmatrix.mat4;
import com.dbooker.livreopengl.glmatrix.vec3;
import com.dbooker.livreopengl.glmatrix.vec4;
import com.dbooker.livreopengl.material.OmniLight;
import com.dbooker.livreopengl.utils.Utils;


public class Scene extends TurnTableScene
{
    // objets de la scène
    private Cow m_Cow;
    private MeshObject m_Grass;
    private OmniLight m_Light0;
    private OmniLight m_Light1;


    /** créer les objets et initialise OpenGL */
    public Scene() throws Exception
    {
        super(true);

        // définir une lampe directionnelle
        m_Light0 = new OmniLight();
        m_Light0.setPosition(vec4.fromValues(1, 1, -1, 0));
        m_Light0.setColor(vec3.fromValues(1.0,1.0,1.0));
        addLight(m_Light0);

        // définir une lampe positionnelle
        m_Light1 = new OmniLight();
        m_Light1.setPosition(vec4.fromValues(-10, 5, 5, 1));
        m_Light1.setColor(vec3.fromValues(150,150,150));
        addLight(m_Light1);

        // créer les objets à dessiner
        m_Cow = new Cow();
        m_Grass = new MeshObjectFromObj("data/models/TerrainSimple", "Terrain.obj", "TerrainHerbe.mtl", 3.0f);

        // configurer les modes de dessin
        glEnable(GL_DEPTH_TEST);
        glEnable(GL_CULL_FACE);

        // couleur du fond
        glClearColor(0.4f, 0.4f, 0.4f, 0.0f);

        // paramètres de la vue
        m_CameraAngleAzimut = 50.0f;
        m_CameraAngleElevation = 10.0f;
        vec3.set(m_CameraPivot, 0.0f, -0.9f, 0.0f);
        m_CameraDistance = 7;
    }


    /**
     * appelée quand la taille de la vue OpenGL change
     * @param width : largeur en nombre de pixels de la fenêtre
     * @param height : hauteur en nombre de pixels de la fenêtre
     */
    @Override
    public void onSurfaceChanged(int width, int height)
    {
        // appeler la méthode de la superclasse
        super.onSurfaceChanged(width, height);

        // matrice de projection (champ de vision)
        mat4.perspective(m_Mat4Projection, Utils.radians(18.0), (float)width / height, 0.1f, 20.0f);
    }


    /**
     * dessine les objets de la scène
     * @param mat4Projection : matrice de projection
     * @param mat4ModelView : matrice de vue
     */
    @Override
    public void onDraw(mat4 mat4Projection, mat4 mat4ModelView)
    {
        // dessiner les objets
        m_Grass.onDraw(mat4Projection, mat4ModelView);
        m_Cow.onDraw(mat4Projection, mat4ModelView);
    }


    /**
     * Cette méthode supprime les ressources allouées
     */
    @Override
    public void destroy()
    {
        m_Light1.destroy();
        m_Light0.destroy();
        m_Grass.destroy();
        m_Cow.destroy();
        super.destroy();
    }
}
